File: exam/03/home/excellent_pair.cc
// alternative provided by TA 鄧晉杰
long ExcellentPair (long n, std::vector<long> Nums)
{
  long amount_pairs {};
  Nums.erase
  (
    std::remove_if
    (
      Nums.begin(),
      Nums.end(),
      [] (long number)
      {
        if (number > 77'777'777)
        {
          return true;
        }
        while (number != 0)
        {
          if (number % 10 > 7)
          {
            return true;
          }
          number /= 10;
        }
        return false;
      }
    ),
    Nums.end()
  );
  std::sort(Nums.begin(), Nums.end());
  auto it {Nums.begin()};
  auto rit {std::prev(Nums.end())};
  while (it != rit)
  {
    auto sum {*it + *rit};
    if (sum == 77'777'777)
    {
      amount_pairs += 2;
      ++it;
    }
    else if (sum < 77'777'777)
    {
      ++it;
    }
    else
    {
      --rit;
    }
  }
  return amount_pairs;
}

File: exam/03/home/excellent_pair.py
# alternative provided by TA 鄧晉杰
def ExcellentPair (n, Nums):
    filtered_nums = []
    for num in Nums:
        is_not_filtered = True
        if num <= 77777777:
            num_ = num
            while num_ != 0:
                if num_ % 10 > 7:
                    is_not_filtered = False
                    break
                num_ //= 10
        else:
            is_not_filtered = False
        if is_not_filtered:
            filtered_nums.append(num)
    filtered_nums.sort()
    amount = 0
    i = 0
    j = len(filtered_nums) - 1
    while i < j:
        sum = filtered_nums[i] + filtered_nums[j]
        if sum == 77777777:
            amount += 2
            i += 1
        elif sum < 77777777:
            i += 1
        else:
            j -= 1
    return amount
